aptos-move/block-executor/src/code_cache_global_manager.rs

```diff
@@ -368,28 +402,185 @@ mod test {
+        assert_struct_name_index_map_size_eq(&manager, 0);
+
+        add_struct_identifier(&mut manager, "foo");
+        add_struct_identifier(&mut manager, "bar");
+        add_struct_identifier(&mut manager, "baz");
+        assert_struct_name_index_map_size_eq(&manager, 3);
+        manager
+            .module_cache
+            .insert(0, mock_verified_code(0, MockExtension::new(8)));
+        manager
+            .module_cache
+            .insert(1, mock_verified_code(1, MockExtension::new(8)));
+        assert_eq!(manager.module_cache.num_modules(), 2);
+        assert_eq!(manager.module_cache.size_in_bytes(), 16);
+
+        // Case 5: Type cache is too large.
+        let metadata_6 = TransactionSliceMetadata::block_from_u64(6, 5);
+        assert!(metadata_6.is_immediately_after(&metadata_5));
+
+        assert_ok!(manager.check_ready(AptosEnvironment::new(&state_view), &config, metadata_6));
+        assert_eq!(manager.transaction_slice_metadata, metadata_6);
+        assert!(manager.environment.is_some());
+        assert_eq!(manager.module_cache.num_modules(), 0);
+        assert_struct_name_index_map_size_eq(&manager, 0);
+    }
+
+    #[test]
+    fn test_try_lock_inner() {
+        let state_view = Arc::new(MockStateView::empty());
+        let config = Arc::new(BlockExecutorModuleCacheLocalConfig::default());
+        let manager = Arc::new(AptosModuleCacheManager::new());
 
+        // Single threaded execution. Locking succeeds.
         let metadata_1 = TransactionSliceMetadata::block_from_u64(0, 1);
-        assert_ok!(manager.try_lock(&state_view, &config, metadata_1));
-        assert_eq!(manager.inner.lock().transaction_slice_metadata, metadata_1);
+        let guard = assert_ok!(manager.try_lock(&MockStateView::empty(), &config, metadata_1));
+        assert!(matches!(guard, AptosModuleCacheManagerGuard::Guard { .. }));
 
+        // Multiple threads try to lock concurrently. Only one succeeds.
         let metadata_2 = TransactionSliceMetadata::block_from_u64(1, 2);
-        assert_ok!(manager.try_lock(&state_view, &config, metadata_2));
-        assert_eq!(manager.inner.lock().transaction_slice_metadata, metadata_2);
-    }
+        let mut handles = Vec::with_capacity(4);
+        for _ in 0..8 {
+            let handle = std::thread::spawn({
+                let manager = manager.clone();
+                let state_view = state_view.clone();
+                let config = config.clone();
+                move || {
+                    let guard =
+                        assert_ok!(manager.try_lock_inner(&state_view, &config, metadata_2));
+                    matches!(guard, AptosModuleCacheManagerGuard::Guard { .. })
+                }
+            });
+            handles.push(handle);
+        }
 
-    // TODO(loader_v2): Add more unit tests like with previous commits.
+        let mut num_guards = 0;
+        let mut num_nones = 0;
+        for handle in handles {
+            if handle.join().unwrap() {
```

Comment:
hm, one of them can call try_lock after you've joined the one that previously succeeded, right? so the test can flake? If I'm not missing anything, then one quick fix would be to add an atomic and increment after try lock, and only join when they all increment it.

Reply:
Indeed.... good catch

Reply:
I think it will always be flaky? the scheduling can always do: thread 1 fully completes, then thread 2 fully completes. I added sleeping to make sure those run concurrently in practice - should be ok?

Reply:
I meant you can share more state to introduce extra barrier after try_lock and before they return, say increment a counter and keep loading until the value is 8 (so they all called try_lock already), then they can clear it and return but the interleavings is restricted, and tests try_lock from different threads.

Reply:
done